src/main/java/org/spongepowered/api/event/CauseStackManager.java

```diff
@@ -0,0 +1,98 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.spongepowered.api.event;
+
+import org.spongepowered.api.event.cause.Cause;
+import org.spongepowered.api.event.cause.EventContext;
+import org.spongepowered.api.event.cause.EventContextKeys;
+
+import java.util.Optional;
+
+/**
+ * Provides an interface into the system managing the cause and contextual
+ * information for events being thrown. This allows plugins to attach more
+ * contextual information to events which may be triggered from their api
+ * interactions.
+ * 
+ * <p>For example a plugin may wish to push a player into the cause stack before
+ * spawning an entity to expose to any plugin listening for SpawnEntityEvents
+ * that the player is the reason that the entity was spawned.</p>
+ * 
+ * <p>This system will automatically handle associating a plugin with actions
+ * performed inside of event listeners, command executors, and scheduled
+ * tasks.</p>
+ */
+public interface CauseStackManager {
+
+    /**
+     * Creates a new {@link Cause} object from the current cause stack.
+     * 
+     * @return A new cause of the current stack.
+     */
+    Cause createCauseFromCurrentState();
+
+    /**
+     * Gets an {@link EventContext} object on the current contextual
+     * information.
+     * 
+     * @return The current event context
+     */
+    EventContext getCurrentContext();
+
+    /**
+     * Pushes an object to the current cause stack which will associate it with
+     * all events through from api actions until it is popped off again.
+     * 
+     * @param obj The object to push to the stack
+     */
+    void pushCause(Object obj);
+
+    /**
+     * Pops the most recently pushed cause object off of the stack and returns
+     * it.
+     * 
+     * @return The last pushed object
+     */
+    Object popCause();
+
+    /**
+     * Adds the given object to the current context under the given key.
+     * 
+     * @param key The context key
+     * @param value The object
+     * @see EventContextKeys
+     */
+    void addContext(String key, Object value);
```

Comment:
If we wanted registered EventContextKeys then we would want a bit more type guarentees potentially adding more keys than what we have now. So its a trade off do we want more explicit typing or more general keys so you don't have to check multiple keys.